### PR TITLE
buildpack dependency fix

### PR DESCRIPTION
**Pull request by noroutine**
There is an apparent dependency issue with yarn buildpack

On a simple repo with yarn.lock, it fails and yarn.lock is ignored. This leads to another buildpack group (with npm) to take over the build.
Result is that whether or not yarn.lock is present, application is always built using npm.

Some pack output demonstrating the issue:

```
✦ ❯❯❯ pack build -v small-yarn -b heroku/nodejs-engine -b heroku/nodejs-yarn
Builder cr.nrtn.dev/heroku/buildpacks:b8c5d04e is trusted
Pulling image cr.nrtn.dev/heroku/buildpacks:b8c5d04e
b8c5d04e: Pulling from heroku/buildpacks
Digest: sha256:79b3131b2f67bce8030fe1824574e4d37d24a7da365233681b65fbc904f2baf4
Status: Image is up to date for cr.nrtn.dev/heroku/buildpacks:b8c5d04e
Selected run image cr.nrtn.dev/infra/pack:20
Pulling image cr.nrtn.dev/infra/pack:20
20: Pulling from infra/pack
Digest: sha256:f774b9e6c779c6e7e168ae5b01eb8f78f393dcc8d02ba88eb22e2e5697f31a5e
Status: Image is up to date for cr.nrtn.dev/infra/pack:20
Setting custom order
Creating builder with the following buildpacks:
-> heroku/procfile@0.6.2
-> heroku/go@0.3.1
-> heroku/nodejs@0.3.5
-> heroku/nodejs-engine@0.7.4
-> heroku/nodejs-npm@0.4.4
-> heroku/nodejs-typescript@0.2.3
-> heroku/nodejs-yarn@0.1.4
-> heroku/procfile@0.6.2
-> heroku/nodejs-function@0.5.7
-> heroku/nodejs-npm@0.4.4
-> heroku/nodejs-typescript@0.2.3
-> heroku/nodejs-engine@0.7.4
-> heroku/nodejs-function-invoker@0.1.6
-> noroutine/static@0.0.3
Using build cache volume pack-cache-library_small-yarn_latest-7623b5239096.build
Running the creator on OS linux with:
Container Settings:
  Args: /cnb/lifecycle/creator -daemon -launch-cache /launch-cache -log-level debug -cache-dir /cache -run-image cr.nrtn.dev/infra/pack:20 -process-type web small-yarn
  System Envs: CNB_PLATFORM_API=0.4
  Image: pack.local/builder/6e617364706f77646c74:latest
  User: root
  Labels: map[author:pack]
Host Settings:
  Binds: pack-cache-library_small-yarn_latest-7623b5239096.build:/cache /var/run/docker.sock:/var/run/docker.sock pack-cache-library_small-yarn_latest-7623b5239096.launch:/launch-cache pack-layers-fmxwuijyaj:/layers pack-app-qbkxereudo:/workspace
  Network Mode:
===> DETECTING
======== Results ========
pass: heroku/nodejs-engine@0.7.4
pass: heroku/nodejs-yarn@0.1.4
Resolving plan... (try #1)
fail: heroku/nodejs-engine@0.7.4 provides unused node
ERROR: No buildpack groups passed detection.
ERROR: Please check that you are running against the correct path.
ERROR: failed to detect: no buildpacks participating
ERROR: failed to build: executing lifecycle: failed with status code: 100
```

Without specific buildpacks and using full set, the group with npm takes over after this issue.

This pull request adds plan dependency on node similar to npm buildpack, that allows yarn to succeed in detection

### CODE REVIEW

**Comment by sawmurai** [comment]
Have you found a workaround for the issue until this gets merged? We are experiencing the same problem, even while just using the heroku/nodejs buildpack: Its using npm instead of yarn.

**Comment by noroutine** [comment]
Hi @sawmurai
I use heroku stacks on my premises, so it was a bit easier to just fix it.
Without modification of buildpacks i didn't find any sensible way to trigger heroku stacks use yarn builder i'm afraid

But I can imagine you can call yarn workflows from npm scripts explicitly, and that way just bypass heroku logic

**Review by joshwlewis — APPROVED** [review]
This looks great. Thanks.

File: buildpacks/yarn/CHANGELOG.md [hunk 1]
```diff
@@ -7,6 +7,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ## [0.1.5] 2021/06/17
 ### Fixed
 - Empty cache builds no longer fail with a `PREV_NODE_VERSION ` unbound variable error ([#86](https://github.com/heroku/buildpacks-node/pull/86))
+- yarn buildpack consumes dependency on node during plan resolution
```
Comment (joshwlewis) [inline buildpacks/yarn/CHANGELOG.md:10]: Would you mind moving this line to the `Unreleased` section? We'll tag this line to the appropriate version when we release it.

**Comment by schneems** [comment]
It looks like shell format checking is failing for the PR due to the newline:

```
$ brew install shfmt
$ shfmt -f . | grep -v ^test/ | grep -v '_shpec.sh$' | xargs shfmt -w
$ git diff
diff --git a/buildpacks/yarn/lib/detect.sh b/buildpacks/yarn/lib/detect.sh
index 8e40dd1..2522bc3 100644
--- a/buildpacks/yarn/lib/detect.sh
+++ b/buildpacks/yarn/lib/detect.sh
@@ -17,4 +17,4 @@ write_to_build_plan() {
        [[requires]]
        name = "node"
 EOF
-}
\ No newline at end of file
+}
```

The output took me a second to realize what was going on:

```
--- buildpacks/yarn/lib/detect.sh.orig
+++ buildpacks/yarn/lib/detect.sh
```

Isn't super helpful, maybe we could have it write and generate a diff on failure in the future?

**Comment by noroutine** [comment]
I figured out that lined ending, thank you, but i don't get what's wrong with changelog now. Diff on my end shows it's in the list 🤔

**Comment by noroutine** [comment]
Oh I see, checkout action is not pulling main branch ref, but there is a diff against it. Seems like related to recent action update here https://github.com/heroku/buildpacks-nodejs/blob/main/.github/workflows/check_changelog.yml#L16

Looks like in v2 they do not checkout main head ref :(

**Comment by noroutine** [comment]
I've rebased against latest main :)

**Comment by joshwlewis** [comment]
Thanks @noroutine -- I've been pushing each of your commits so they trigger CI. I've got my fingers crossed for this commit =P.

**Comment by edmorley** [comment]
@joshwlewis @noroutine Hi! The Yarn cutlass test can now be updated to remove the manually specified yarn buildpack, and rely just on the main `nodejs` meta-buildpack:
https://github.com/heroku/buildpacks-nodejs/blob/d727ef98c3c4b43d3d862541dc43e4490e8a460c/test/specs/node/cutlass/basic_spec.rb#L26-L28

**Comment by noroutine** [comment]
Hey @edmorley 
Updated :) Catch another PR for that https://github.com/heroku/buildpacks-nodejs/pull/109

**Comment by edmorley** [comment]
@noroutine Thank you! I'll leave to Josh to review (new Node.js owner).

**Pull request merged**